### PR TITLE
Add DLPack bool support

### LINKED ISSUE #67081: DLPack no longer works on Boolean tensors after 1.10+
## 🐛 Bug

`torch.utils.dlpack.to_dlpack` no longer works for Boolean Tensor.

## To Reproduce

```python
torch.utils.dlpack.to_dlpack(torch.BoolTensor([False, True]))   # Bool type is not supported by dlpack
```

## Expected behavior

Should work.

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```

 - PyTorch Version (e.g., 1.0): 1.10 RC and nightly
 - OS (e.g., Linux): Linux
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Build command you used (if compiling from source):
 - Python version: 3.8.3
 - CUDA/cuDNN version:
 - GPU models and configuration:
 - Any other relevant information:

## Additional context

According to https://github.com/dmlc/dgl/pull/3406 the error is raised after https://github.com/pytorch/pytorch/pull/57110.  I was wondering what the reason was behind removing boolean support?

This is blocking DGL's patch release with PyTorch 1.10 since we rely on DLPack to interact with PyTorch tensors in multiple places.

+@jermainewang @VoVAllen

cc @brianjo @mruberry @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser

### LINKED ISSUE #94463: DLPack doesn't Support Boolean Tensor
### 🐛 Describe the bug

Latest DLPack already has bool dtype support in DLDataTypeCode (https://github.com/dmlc/dlpack/blob/ca4d00ad3e2e0f410eeab3264d21b8a39397f362/include/dlpack/dlpack.h#L157). Please consider updating the dlpack.h header file to latest and adding boolean tensor support to both to_dlpack and from_dlpack.

To reproduce:
`torch.utils.dlpack.to_dlpack(torch.BoolTensor([False, True]))   # Bool type is not supported by dlpack`

### Versions

Collecting environment information...
PyTorch version: 2.0.0.dev20230202+cu116
Is debug build: False
CUDA used to build PyTorch: 11.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~18.04) 9.4.0
Clang version: 15.0.0-++20220616064206+d0e60b6d7ec9-1~exp1~20220616184301.556
CMake version: version 3.24.1
Libc version: glibc-2.35

Python version: 3.8.5 (default, Sep  4 2020, 07:30:14)  [GCC 7.3.0] (64-bit runtime)
Python platform: Linux-4.15.0-187-generic-x86_64-with-glibc2.10
Is CUDA available: True
CUDA runtime version: 11.6.124
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: Tesla V100-DGXS-32GB
GPU 1: Tesla V100-DGXS-32GB
GPU 2: Tesla V100-DGXS-32GB
GPU 3: Tesla V100-DGXS-32GB

Nvidia driver version: 510.47.03
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.4.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.4.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.4.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.4.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.4.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.4.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.4.1
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn.so.8.2.1
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.2.1
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.2.1
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.2.1
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.2.1
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.2.1
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.2.1
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              40
On-line CPU(s) list: 0-39
Thread(s) per core:  2
Core(s) per socket:  20
Socket(s):           1
NUMA node(s):        1
Vendor ID:           GenuineIntel
CPU family:          6
Model:               79
Model name:          Intel(R) Xeon(R) CPU E5-2698 v4 @ 2.20GHz
Stepping:            1
CPU MHz:             1198.138
CPU max MHz:         3600.0000
CPU min MHz:         1200.0000
BogoMIPS:            4397.44
Virtualization:      VT-x
L1d cache:           32K
L1i cache:           32K
L2 cache:            256K
L3 cache:            51200K
NUMA node0 CPU(s):   0-39
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 invpcid_single pti intel_ppin ssbd ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm rdt_a rdseed adx smap intel_pt xsaveopt cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts md_clear flush_l1d

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.21.6
[pip3] numpydoc==1.1.0
[pip3] pytorch-triton==2.0.0+0d7e753227
[pip3] torch==2.0.0.dev20230202+cu116
[pip3] torchaudio==2.0.0.dev20230201+cu116
[pip3] torchmetrics==0.7.3
[pip3] torchvision==0.15.0.dev20230201+cu116
[conda] blas                      1.0                         mkl  
[conda] magma-cuda110             2.5.2                         1    pytorch
[conda] mkl                       2020.2                      256  
[conda] mkl-include               2020.2                      256  
[conda] mkl-service               2.3.0            py38he904b0f_0  
[conda] mkl_fft                   1.2.0            py38h23d657b_0  
[conda] mkl_random                1.1.1            py38h0573a6f_0  
[conda] numpy                     1.21.6                   pypi_0    pypi
[conda] numpydoc                  1.1.0              pyhd3eb1b0_1  
[conda] pytorch-triton            2.0.0+0d7e753227          pypi_0    pypi
[conda] torch                     1.13.0.dev20220724+cu113          pypi_0    pypi
[conda] torchaudio                2.0.0.dev20230201+cu116          pypi_0    pypi
[conda] torchmetrics              0.7.3                    pypi_0    pypi
[conda] torchvision               0.15.0.dev20230201+cu116          pypi_0    pypi

### PR DESCRIPTION
**Pull request by Tabrizian**
Fixes #94463
Fixes https://github.com/pytorch/pytorch/issues/67081

- [X] Update DLPack header file
- [X] Add testing for DLPack boolean
- [X] Add boolean support to PyTorch's DLPack support

### CODE REVIEW

File: aten/src/ATen/dlpack.h
Comment (IvanYashchuk) [inline aten/src/ATen/dlpack.h:1]: This is just an updated copy of https://github.com/dmlc/dlpack/blob/main/include/dlpack/dlpack.h, right?

Comment (ezyang) [reply aten/src/ATen/dlpack.h:1]: Yes

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
Need CLA. We should also disable lint on dlpack header

**Comment by Tabrizian** [comment]
@ezyang Thanks for the review. I've requested to be included in the corporate (NVIDIA) CLA.

File: aten/src/ATen/dlpack.h
Comment (Tabrizian) [reply aten/src/ATen/dlpack.h:1]: I've rolled back to v0.8.

**Comment by Tabrizian** [comment]
@ezyang Do I need to add release notes for this PR or I can just use "topic: not user facing"?

**Comment by Tabrizian** [comment]
@ezyang The failures seem to be unrelated to my change. Let me know if there is anything missing from my end.

**Pull request closed without merging**